src/commands/hld/reconcile.ts

```diff
@@ -0,0 +1,182 @@
+
+    // Fab add is idempotent.
+    // mkdir -p does not fail if ${pathBase} does not exist.
+    const createSvcInHldCommand = `cd ${absRepositoryInHldPath} && mkdir -p ${pathBase} config && fab add ${pathBase} --source ./${pathBase} --method local`;
+
+    await execAndLog(createSvcInHldCommand);
+
+    // No rings
+    if (!managedRings) {
+      continue;
+    }
+
+    // Create ring components.
+    const svcPathInHld = path.join(absRepositoryInHldPath, pathBase);
+    for (const ring of Object.keys(managedRings)) {
+      const ringPathInHld = path.join(svcPathInHld, ring);
+
+      // If the ring component already exists, we do not attempt to create it.
+      if (
+        shelljs.test("-e", ringPathInHld) &&
+        shelljs.test("-d", ringPathInHld)
+      ) {
+        logger.info(
+          `Ring component: ${ring} already exists, skipping ring generation.`
+        );
+        continue;
+      }
+
+      // Otherwise, create the ring in the service.
+      const createRingInSvcCommand = `cd ${svcPathInHld} && mkdir -p ${ring} config && fab add ${ring} --source ./${ring} --method local`;
+
+      await execAndLog(createRingInSvcCommand);
+
+      let addHelmChartCommand = "";
+      if (helmConfig.helm.chart.method === "git") {
+        // TODO: git sha
+        addHelmChartCommand = `fab add chart --source ${helmConfig.helm.chart.git} --path ${helmConfig.helm.chart.path}`;
+      } else {
+        addHelmChartCommand = `fab add chart --source ${helmConfig.helm.chart.repository} --path ${helmConfig.helm.chart.chart}`;
+      }
+
+      await execAndLog(addHelmChartCommand);
+
+      // Create config directory, crate static manifest directory.
+      const createConfigAndStaticComponentCommand = `cd ${ringPathInHld} && mkdir -p config static && fab add static --source ./static --method local --type static`;
+
+      await execAndLog(createConfigAndStaticComponentCommand);
+
+      // Create Ingress Route.
+      const staticComponentPathInRing = path.join(ringPathInHld, "static");
+      const ingressRoutePathInStaticComponent = path.join(
+        staticComponentPathInRing,
+        "ingress-route.yaml"
+      );
+
+      // TODO: figure out a way to grab the port from _somewhere_; store in bedrock.yaml?
+      const ingressRoute = TraefikIngressRoute(serviceName, ring, 8000);
+      const routeYaml = yaml.safeDump(ingressRoute, {
+        lineWidth: Number.MAX_SAFE_INTEGER
+      });
+
+      logger.info(
+        `Writing IngressRoute YAML to ${ingressRoutePathInStaticComponent}`
+      );
+      writeFileSync(ingressRoutePathInStaticComponent, routeYaml);
+    }
+  }
+};
+
+/**
+ * Runs a command via `exec` and captures the results, logging the command
+ * that will be run, and the results of that command.
+ *
+ * @param commandToRun String version of the command that must be run
+ * @throws An Error containing the logs captured from stderr
+ */
+const execAndLog = async (commandToRun: string) => {
```

Comment:
Perhaps we'll want this in a lib later. A bit confusing with this reference and us having an `exec` library too.

Reply:
fair